### PR TITLE
CBG-2686: fix for dcp resync flake

### PR DESCRIPTION
**Pull request by gregns1**
CBG-2686

We were asseerting on docs prcessed on dcp resync. I think this was causing some flaking results. Due to nature of dcp it's hard to assert on what you see over dcp so I have changed the assertions to be at least the number of docs added in the test case and thus have to change assertion on sync function runs stats too. This will behind the `isDCPResync` flag and the original, assertion will remain there for non dcp runs. 
I have then added a check in the code to get status doc from the cluster. We were originally just setting sttaus to stopped if heartbeat isnt found. It is entirely possible that the status is completed and thus the heartbeat doc won't be there. So if the status is completed we cant it to remain this way indicating the background process has been successful. 

* Took a slightly different apprioach as it still flaked on git hub actions once. I think this now fixes the issue, I have re run jenkins github actions a few times now with no flake for this test. 

## Pre-review checklist
- [x] Removed debug logging (`fmt.Print`, `log.Print`, ...)
- [x] Logging sensitive data? Make sure it's tagged (e.g. `base.UD(docID)`, `base.MD(dbName)`)
- [x] Updated relevant information in the API specifications (such as endpoint descriptions, schemas, ...) in `docs/api`

## [Integration Tests](https://jenkins.sgwdev.com/job/SyncGateway-Integration/build?delay=0sec)
- [x] `GSI=true,xattrs=true` https://jenkins.sgwdev.com/job/SyncGateway-Integration/1689/ (Run dcp resync 200 times)
- [x] `GSI=true,xattrs=true` https://jenkins.sgwdev.com/job/SyncGateway-Integration/1690/ (Run non dcp resync 200 times against server)

### CODE REVIEW

File: rest/sync_fn_test.go [hunk 1]
```diff
@@ -603,19 +603,21 @@ func TestResync(t *testing.T) {
 			})
 			assert.NoError(t, err)
 
-			assert.Equal(t, testCase.expectedSyncFnRuns, int(rt.GetDatabase().DbStats.Database().SyncFunctionCount.Value()))
-
 			if !isDCPResync {
 				var queryName string
+				assert.Equal(t, testCase.expectedSyncFnRuns, int(rt.GetDatabase().DbStats.Database().SyncFunctionCount.Value()))
```
Comment (torcolvin) [inline rest/sync_fn_test.go]: These assertions are all dead - it is _always_ DCP resync unless `UseQueryBasedResyncManager` is set.

File: rest/sync_fn_test.go [hunk 2]
```diff
@@ -512,7 +512,10 @@ func TestDBOfflineSingleResync(t *testing.T) {
 	assert.Equal(t, int64(2000), rt.GetDatabase().DbStats.Database().SyncFunctionCount.Value())
 }
 
-func TestResync(t *testing.T) {
+func TestResyncOverDCP(t *testing.T) {
+	if base.UnitTestUrlIsWalrus() {
+		t.Skip("This test doesn't works with walrus")
```
Comment (torcolvin) [inline rest/sync_fn_test.go]:
```suggestion
                t.Skip("DCP requires gocb and CBS")
```

File: db/background_mgr.go [hunk 3]
```diff
@@ -307,10 +307,12 @@ func (b *BackgroundManager) getStatusFromCluster() ([]byte, error) {
 		_, _, err = b.clusterAwareOptions.metadataStore.GetRaw(b.clusterAwareOptions.HeartbeatDocID())
 		if err != nil {
 			if base.IsDocNotFoundError(err) {
-				clusterStatus["status"] = BackgroundProcessStateStopped
-				status, err = base.JSONMarshal(clusterStatus)
-				if err != nil {
-					return nil, err
+				if clusterStatus["status"] != BackgroundProcessStateCompleted {
```
Comment (torcolvin) [inline db/background_mgr.go]:
```suggestion
				if clusterState != BackgroundProcessStateCompleted
```

File: rest/sync_fn_test.go [hunk 4]
```diff
@@ -603,21 +601,8 @@ func TestResync(t *testing.T) {
 			})
 			assert.NoError(t, err)
```
Comment (torcolvin) [inline rest/sync_fn_test.go]:
```suggestion
			require.NoError(t, err)
```

**Review by torcolvin — COMMENTED** [review]
Just a few small nitpicks for readability in the future.

**Comment by torcolvin** [comment]
I am OK with this change and if it fixes GH actions run a sufficient number of times that's OK. I'd like to see a test that works solely on Background Manager where it starts/stops the goroutines quickly since this fix is BackgroundManager specific.

But if that's hard to reproduce, I'm OK with merging this.

**Comment by gregns1** [comment]
> I am OK with this change and if it fixes GH actions run a sufficient number of times that's OK. I'd like to see a test that works solely on Background Manager where it starts/stops the goroutines quickly since this fix is BackgroundManager specific.
> 
> But if that's hard to reproduce, I'm OK with merging this.

Added temporary change to run the test 100 times in gh actions and no failiure. Also re ran the tests in the jenkins bit in actions and no failiures. Think this fixes the issue but will reopen the ticket if we see it again.

**Review by torcolvin — APPROVED** [review]

**Pull request merged**